src/extension.ts

```diff
@@ -535,7 +553,7 @@ export class Extension implements RunHooks {
     if (isDebug) {
       await model.debugTests(items, testListener, testRun.token);
     } else {
-      await this._traceViewer.willRunTests(model.config);
+      await this._traceViewer.willRunTests(await this._extendTestConfig(model));
```

Comment:
It seems like moving the trace viewer instantiation into PlaywrightTestCLI / PlaywrightTestServer would be a meaningful refactoring before this change. Something like

```
if (this._traceViewer)
  this._traceViewer.close();
this._traceViewer = await model.showTraceViewerIfNeeded(); // delegates to PlaywrightTest{CLI/Server}
await model.runTests(items, testListener, testRun.token);
```

Models will run in-order sequentially, so before each model run (that corresponds to a specific test sever), you'll be opening trace viewer using that exact test server.

Reply:
Ok, understood. That way, I can even pass the `model.config` to traceViewer constructor, instead of passing it in each call.

I was also considering splitting `TraceViewer` into two implementations: one with the legacy logic and another with the embedded logic. I think it simplifies the code (right now, I find it a little messy). And with your new proposal, it will be easy to ensure that we start the right TraceViewer implementation each time.

Reply:
7644d39 contains your proposal and the trace viewer split. Due to the nature of current `TraceViewer` (I renamed it to `SpawnTraceViewer`), which is shared among testModels, its lyfecycle is still managed in `extension.ts`.

Each model can provide an optional `EmbededTraceViewer` via `PlaywrightTestServer`. Then when we need a trace viewer we check the settings and if the model supports embedded mode, otherwise we fallback to `SpawnTraceViewer`.

This way, we can keep the current use case of multiple test models using the same `SpawnTraceViewer` without needed to close and open it each time.

Reply:
I'm testing scenarios with multiple test configs. When running a test set from multiple test configs, it iterates over each test config, as you said, but it will close embedded trace viewer too many times and unnecessarily. That doesn't happen when we just select test items with traces because it always uses the selected test model in those cases.

I think we could do the same on `Extension._runTest`: instead of using each test config to pick the corresponding trace viewer, we always use the selected test config one.